python/dgl/heterograph.py

```diff
@@ -0,0 +1,1387 @@
+            metagraph,
+            number_of_nodes_by_type,
+            edge_connections_by_type):
+        super(DGLBaseHeteroGraph, self).__init__()
+
+    def __getitem__(self, key):
+        """Returns a view on the heterogeneous graph with given node/edge
+        type:
+
+        * If ``key`` is a str, it returns a heterogeneous subgraph induced
+          from nodes of type ``key``.
+        * If ``key`` is a pair of str (type_A, type_B), it returns a
+          heterogeneous subgraph induced from the union of both node types.
+        * If ``key`` is a triplet of str
+
+              (src_type_name, dst_type_name, edge_type_name)
+
+          It returns a heterogeneous subgraph induced from the edges with
+          source type name ``src_type_name``, destination type name
+          ``dst_type_name``, and edge type name ``edge_type_name``.
+
+        The view would share the frames with the parent graph; any
+        modifications on one's frames would reflect on the other.
+
+        Note that the subgraph itself is not materialized until someone
+        queries the subgraph structure.  This implies that calling computation
+        methods such as
+
+            g['user'].update_all(...)
+
+        would not create a subgraph of users.
+
+        Parameters
+        ----------
+        key : str or tuple
+            See above
+
+        Returns
+        -------
+        DGLBaseHeteroGraphView
+            The induced subgraph view.
+        """
+        pass
+
+
+class DGLBaseHeteroGraphView(object):
+    """View on a heterogeneous graph, constructed from
+    DGLBaseHeteroGraph.__getitem__().
+
+    It is semantically the same as a subgraph, except that
+
+    * The subgraph itself is not materialized until the user explicitly
+    queries the subgraph structure (e.g. calling ``in_edges``, but not
+    ``update_all``).
+    """
+    pass
+
+
+class DGLHeteroGraph(DGLBaseHeteroGraph):
+    """Base heterogeneous graph class.
+
+    The graph stores nodes, edges and also their (type-specific) features.
+
+    Heterogeneous graphs are by default multigraphs.
+
+    Parameters
+    ----------
+    metagraph, number_of_nodes_by_type, edge_connections_by_type :
+        See DGLBaseHeteroGraph
+    node_frame : dict[str, FrameRef], optional
+        Node feature storage per type
+    edge_frame : dict[str, FrameRef], optional
+        Edge feature storage per type
+    readonly : bool, optional
+        Whether the graph structure is read-only (default: False)
+    """
+    # pylint: disable=unused-argument
+    def __init__(
+            self,
+            metagraph,
+            number_of_nodes_by_type,
+            edge_connections_by_type,
+            node_frame=None,
+            edge_frame=None,
+            readonly=False):
+        super(DGLHeteroGraph, self).__init__(
+            metagraph, number_of_nodes_by_type, edge_connections_by_type)
+
+    # TODO: REVIEW
+    def add_nodes(self, num, node_type, data=None):
```

Comment:
Should these interfaces be declared in `DGLBaseHeteroGraph `?

Reply:
I don't know.  I followed `DGLBaseGraph` which doesn't have `add_nodes` method.

Reply:
I feel it's more appropriate to have all of these methods in `DGLBaseHeteroGraph`. Otherwise, I don't which class `DGLBipartiteGraph` should inherit from.